aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -731,7 +731,78 @@ Tensor dense_to_sparse_csr(const Tensor& self) {
 }
 
 Tensor dense_to_sparse_csc(const Tensor& self) {
-  return self.to_sparse().to_sparse_csc();
+  auto n_batch_dim = self.dim() - 2;
+  auto values = self;
+  auto not_zero_mask = self != 0;
+
+  if (n_batch_dim > 0) {
+    if (n_batch_dim > 1) {
+      not_zero_mask = not_zero_mask.flatten(0, n_batch_dim - 1);
+      values = values.flatten(0, n_batch_dim - 1);
+    }
+
+    TORCH_CHECK(
+        not_zero_mask.size(0) > 0,
+        "to_sparse_csc: Expected product of batch dimensions to be non-zero.");
+
+    // If the input is n-d we require that nnz is the same for all
+    // batches.
+    auto not_zero_mask_0 = not_zero_mask.select(0, 0);
+    auto nse_per_batch = not_zero_mask_0.sum().repeat(not_zero_mask.size(0));
+    TORCH_CHECK(
+        not_zero_mask.sum({-2, -1}).equal(nse_per_batch),
+        "Expect the same number of specified elements per batch.");
+
+    // We need to join batches into a matrix increasing the number of cols.
+    // This allows us to create indices for col compressed matrix and de-batch
+    // them later. Avoiding a kernel start for each batch.
+    values = values.transpose(0, 1).flatten(1, 2);
+    not_zero_mask = not_zero_mask.transpose(0, 1).flatten(1, 2);
+  }
```

Comment:
wdyt of factoring the invariant checks (#738-760) into a helper that's called from both `dense_to_sparse_csr` and `dense_to_sparse_csc`? getting common stuff out of the builder code below is more subtle but I think this acts only on the dense input and is identical (mod messaging), if I'm reading it right

Reply:
738-754 could be factored out and shared among `dense_to_sparse_{csr,bsr,csc,bsc}` the last two lines differ subtly. `csc/bsc` require a transpose and the flatten is shifted over.

Reply:
@bhosmer FYI: I refactored the handling of batch-dims on the setup (including checks) and the output side into two helpers. They have been used across the stack so the changes are on the base PR, see [this diff](https://github.com/pytorch/pytorch/pull/80781/files#diff-d4df180433a09071e891d552426911c227b30ae9b8a8e56da31046e7ecb1afbeR17-R108)